src/main/java/spoon/support/compiler/jdt/ReferenceBuilder.java

```diff
@@ -260,12 +260,13 @@ private boolean isTypeArgumentExplicit(TypeReference[][] typeArguments) {
 	 * Builds a type reference from a qualified name when a type specified in the name isn't available.
 	 *
 	 * @param tokens        Qualified name.
+	 * @param receiverNode  AST node of last type in the qualified name.
 	 * @param receiverType  Last type in the qualified name.
 	 * @param enclosingType Enclosing type of the type name.
 	 * @param listener      Listener to know if we must build the type reference.
 	 * @return a type reference.
 	 */
-	<T> CtTypeReference<T> getQualifiedTypeReference(char[][] tokens, TypeBinding receiverType, ReferenceBinding enclosingType, JDTTreeBuilder.OnAccessListener listener) {
+	<T> CtTypeReference<T> getQualifiedTypeReference(char[][] tokens, ASTNode receiverNode, TypeBinding receiverType, ReferenceBinding enclosingType, JDTTreeBuilder.OnAccessListener listener) {
```

Comment:
Adding a new parameter here is fine from my side, because this isn't real public api. Any other opinions?

Reply:
I needed access to AST node to build the source position so for now, I just passed it as a parameter. I have opened an issue https://github.com/INRIA/spoon/issues/4297 to find a more generic solution to this but that would require how `CtTypeReference` are built and inserted into the model. This PR is kind of like a workaround but it fixes a very important issue in Sorald.

I will come back to https://github.com/INRIA/spoon/issues/4297 once I am done dealing with some important issues in sorald.

Reply:
It's generally fine to add parameters to these internal methods (anything that isn't public, really), but here I think it's not the right solution. The added parameter is used in exactly the same way in two different places, which indicates to me that it should be used at the call site of the method rather than being passed in. So I think the code changes to this method should be reverted.

I think the place to solve this particular problem is in `JDTTreeBuilder.visit(QualifiedTypeReference, BlockScope)`, and change the very end of the method like so:

```diff
-		context.enter(factory.Code().createTypeAccessWithoutCloningReference(references.buildTypeReference(qualifiedTypeReference, scope)), qualifiedTypeReference);
+		CtTypeReference<?> typeRef = references.buildTypeReference(qualifiedTypeReference, scope);
+		if (typeRef != null) {
+			typeRef.setPosition(position.buildPositionCtElement(typeRef, qualifiedTypeReference));
+		}
+		context.enter(factory.Code().createTypeAccessWithoutCloningReference(typeRef), qualifiedTypeReference);
```